### PR TITLE
[stable/rabbitmq] Added volume permissions

### LINKED ISSUE #11926: [stable/rabbitmq] Permission denied on NFS storage class with securityContext
**Version of Helm and Kubernetes**:
Client: &version.Version{SemVer:"v2.12.1", GitCommit:"02a47c7249b1fc6d8fd3b94e6b4babf9d818144e", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.12.1", GitCommit:"02a47c7249b1fc6d8fd3b94e6b4babf9d818144e", GitTreeState:"clean"}

**Which chart**:
stable/rabbitmq-4.3.1

**What happened**:
Permission denied error when rabbitmq is initialized

**What you expected to happen**:
Run normally

**How to reproduce it** (as minimally and precisely as possible):
Use the following ``values.yaml`` with an NFS storage class (default in my use case)

```
## Global Docker image registry
## Please, note that this will override the image registry for all the images, including dependencies, configured to use the global value
##
# global:
#   imageRegistry:

## Bitnami RabbitMQ image version
## ref: https://hub.docker.com/r/bitnami/rabbitmq/tags/
##
image:
  registry: docker.io
  repository: bitnami/rabbitmq
  tag: 3.7.12

  ## set to true if you would like to see extra information on logs
  ## it turns BASH and NAMI debugging in minideb
  ## ref:  https://github.com/bitnami/minideb-extras/#turn-on-bash-debugging
  debug: false

  ## Specify a imagePullPolicy
  ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
  ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
  ##
  pullPolicy: IfNotPresent
  ## Optionally specify an array of imagePullSecrets.
  ## Secrets must be manually created in the namespace.
  ## ref: https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/
  ##
  # pullSecrets:
  #   - myRegistrKeySecretName

## does your cluster have rbac enabled? assume yes by default
rbacEnabled: true

## section of specific values for rabbitmq
rabbitmq:
  ## RabbitMQ application username
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  username: user

  ## RabbitMQ application password
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  # password:
  # existingPasswordSecret: name-of-existing-secret

  ## Erlang cookie to determine whether different nodes are allowed to communicate with each other
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  # erlangCookie:
  # existingErlangSecret: name-of-existing-secret

  ## Node name to cluster with. e.g.: `clusternode@hostname`
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  # rabbitmqClusterNodeName:

  ## RabbitMQ Max File Descriptors
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ## ref: https://www.rabbitmq.com/install-debian.html#kernel-resource-limits
  ##
  ulimitNofiles: '65536'

  ## Plugins to enable
  plugins: |-
      [rabbitmq_management, rabbitmq_peer_discovery_k8s].

  ## Clustering settings
  clustering:
    address_type: ip
    k8s_domain: cluster.local

  ## Configuration file content: required cluster configuration
  ## Do not override unless you know what you are doing. To add more configuration, use `extraConfiguration` instead
  configuration: |-
    ## Clustering
    cluster_formation.peer_discovery_backend  = rabbit_peer_discovery_k8s
    cluster_formation.k8s.host = kubernetes.default.svc.cluster.local
    cluster_formation.node_cleanup.interval = 10
    cluster_formation.node_cleanup.only_log_warning = true
    cluster_partition_handling = autoheal
    # queue master locator
    queue_master_locator=min-masters
    # enable guest user
    loopback_users.guest = false

  ## Configuration file content: extra configuration
  ## Use this instead of `configuration` to add more configuration
  extraConfiguration: |-
    #disk_free_limit.absolute = 50MB

## Kubernetes service type
service:
  type: ClusterIP
  ## Node port
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  nodePort: 5672

  ## Amqp port
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  amqpPort: 5672
  ## Dist port
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  distPort: 25672

  ## RabbitMQ Manager port
  ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
  ##
  managerPort: 15672

# Additional pod labels to apply
podLabels: {}

## Pod Security Context
## ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
##
securityContext:
  enabled: true
  fsGroup: 1001
  runAsUser: 1001

persistence:
  ## this enables PVC templates that will create one per pod
  enabled: true

  ## rabbitmq data Persistent Volume Storage Class
  ## If defined, storageClassName: <storageClass>
  ## If set to "-", storageClassName: "", which disables dynamic provisioning
  ## If undefined (the default) or set to null, no storageClassName spec is
  ##   set, choosing the default provisioner.  (gp2 on AWS, standard on
  ##   GKE, AWS & OpenStack)
  ##
  # storageClass: "-"
  accessMode: ReadWriteOnce

  # If you change this value, you might have to adjust `rabbitmq.diskFreeLimit` as well.
  size: 8Gi

## Configure resource requests and limits
## ref: http://kubernetes.io/docs/user-guide/compute-resources/
##
resources:
  requests:
    memory: 256Mi
    cpu: 100m

## Replica count, set to 3 to provide a default available cluster
replicas: 3

## Pod priority
## https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/
# priorityClassName: ""

## Node labels and tolerations for pod assignment
## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#taints-and-tolerations-beta-feature
nodeSelector:
  beta.kubernetes.io/arch: amd64
tolerations: []
affinity: {}

## annotations for rabbitmq pods
podAnnotations: {}

## Configure the ingress resource that allows you to access the
## Wordpress installation. Set up the URL
## ref: http://kubernetes.io/docs/user-guide/ingress/
##
ingress:
  ## Set to true to enable ingress record generation
  enabled: false

  ## The list of hostnames to be covered with this ingress record.
  ## Most likely this will be just one host, but in the event more hosts are needed, this is an array
  ## hostName: foo.bar.com
  path: /

  ## Set this to true in order to enable TLS on the ingress record
  ## A side effect of this will be that the backend wordpress service will be connected at port 443
  tls: true

  ## If TLS is set to true, you must declare what secret will store the key/certificate for TLS
  tlsSecret: myTlsSecret

  ## Ingress annotations done as key:value pairs
  ## If you're using kube-lego, you will want to add:
  ## kubernetes.io/tls-acme: true
  ##
  ## For a full list of possible ingress annotations, please see
  ## ref: https://github.com/kubernetes/ingress-nginx/blob/master/docs/annotations.md
  ##
  ## If tls is set to true, annotation ingress.kubernetes.io/secure-backends: "true" will automatically be set
  annotations:
  #  kubernetes.io/ingress.class: nginx
  #  kubernetes.io/tls-acme: true

## The following settings are to configure the frequency of the lifeness and readiness probes
livenessProbe:
  enabled: true
  initialDelaySeconds: 120
  timeoutSeconds: 20
  periodSeconds: 30
  failureThreshold: 6
  successThreshold: 1

readinessProbe:
  enabled: true
  initialDelaySeconds: 10
  timeoutSeconds: 20
  periodSeconds: 30
  failureThreshold: 3
  successThreshold: 1

metrics:
  enabled: true
  image:
    registry: docker.io
    repository: kbudde/rabbitmq-exporter
    tag: v0.29.0
    pullPolicy: IfNotPresent
  resources: {}
  annotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "9090"
```

**Anything else we need to know**:
NFS persistent volume is created with root:root permissions and when the following line is executed inside the rabbitmq container:

```
echo $RABBITMQ_ERL_COOKIE > /opt/bitnami/rabbitmq/var/lib/rabbitmq/.erlang.cookie
```
we get a permission denied error

### PR DESCRIPTION
**Pull request by artemistomaras**
#### What this PR does / why we need it:
This pull request adds an option for the **stable/rabbitmq** chart to set the permissions inside the data directory with an init-container when the securityContext and the persistence options are set to ``true``.
 
#### Which issue this PR fixes
  - fixes #11926

#### Checklist
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by artemistomaras** [comment]
I am not sure about the version bump. I leave it as-is unless you tell me otherwise

File: stable/rabbitmq/Chart.yaml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: rabbitmq
-version: 4.3.1
+version: 4.3.2
```
Comment (juan131) [inline stable/rabbitmq/Chart.yaml]: Increase the minor version.

File: stable/rabbitmq/values-production.yaml [hunk 2]
```diff
@@ -96,12 +96,13 @@ service:
   ## Node port
   ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
   ##
-  nodePort: 5672
+  # nodePort: 30672
```
Comment (juan131) [inline stable/rabbitmq/values-production.yaml:107]: Undo this change.

File: stable/rabbitmq/values-production.yaml [hunk 3]
```diff
@@ -96,12 +96,13 @@ service:
...
 
   ## Amqp port
   ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
   ##
-  amqpPort: 5672
+  port: 5672
```
Comment (juan131) [inline stable/rabbitmq/values-production.yaml:112]: Undo this change too.

File: stable/rabbitmq/templates/statefulset.yaml [hunk 4]
```diff
@@ -53,6 +53,20 @@ spec:
 {{ toYaml .Values.tolerations | indent 8 }}
       {{- end }}
       terminationGracePeriodSeconds: 10
+      {{- if and .Values.volumePermissions.enabled .Values.persistence.enabled .Values.securityContext.enabled }}
+      initContainers:
+      - name: volume-permissions
+        image: "{{ template "volumePermissions.image" . }}"
+        imagePullPolicy: {{ default "" .Values.volumePermissions.image.pullPolicy | quote }}
+        command: ["/bin/chown", "-R", "{{ .Values.securityContext.runAsUser }}:{{ .Values.securityContext.fsGroup }}", "/opt/bitnami/rabbitmq/var/lib/rabbitmq/"]
```
Comment (juan131) [inline stable/rabbitmq/templates/statefulset.yaml]: I'd use sth like:

```yaml
        command: ["/bin/chown", "-R", "{{ .Values.securityContext.runAsUser }}:{{ .Values.securityContext.fsGroup }}", "{{ .Values.persistence.path }}"]
```

To do so, you need to create the parameter `persistence.path` and add it to `values.yaml` and `values-production.yaml` files (and README.md)

File: stable/rabbitmq/values-production.yaml [hunk 5]
```diff
@@ -226,3 +227,16 @@ metrics:
...
+## volumePermissions: Change the owner of the persist volume mountpoint to RunAsUser:fsGroup
+##
+volumePermissions:
+  enabled: true
+  image:
+    registry: docker.io
+    repository: bitnami/minideb
+    tag: latest
+    pullPolicy: IfNotPresent
+  resources: {}
```
Comment (juan131) [inline stable/rabbitmq/values-production.yaml:259]: This parameter `volumePermissions.resources` is not documented in the README.md

File: stable/rabbitmq/templates/statefulset.yaml [hunk 6]
```diff
@@ -53,6 +53,20 @@ spec:
...
+        securityContext:
+          runAsUser: 0
+        resources:
+{{ toYaml .Values.volumePermissions.resources | indent 10 }}
+        volumeMounts:
+        - name: data
+          mountPath: /opt/bitnami/rabbitmq/var/lib/rabbitmq/
```
Comment (juan131) [inline stable/rabbitmq/templates/statefulset.yaml]: Again, I'd use:

```yaml
          mountPath: {{ .Values.persistence.path }}
```

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for the PR! Please take a look to my suggestions.

File: stable/rabbitmq/Chart.yaml (hunk 1 above)
Comment (artemistomaras) [reply stable/rabbitmq/Chart.yaml]: Is 4.4.1 reasonable?

File: stable/rabbitmq/values-production.yaml (hunk 2 above)
Comment (artemistomaras) [reply stable/rabbitmq/values-production.yaml:107]: Ok

File: stable/rabbitmq/values-production.yaml (hunk 3 above)
Comment (artemistomaras) [reply stable/rabbitmq/values-production.yaml:112]: Isn't ``amqpPort`` a typo?

I cannot see that reference inside the ``svc-headless.yaml`` nor the ``svc.yaml`` templates.
Also, commenting out the ``port`` value, will leave tjhe port value null:

https://github.com/helm/charts/blob/master/stable/rabbitmq/templates/svc-headless.yaml#L17
https://github.com/helm/charts/blob/master/stable/rabbitmq/templates/svc.yaml#L21

File: stable/rabbitmq/templates/statefulset.yaml (hunk 4 above)
Comment (artemistomaras) [reply stable/rabbitmq/templates/statefulset.yaml]: This makes sense, will do

File: stable/rabbitmq/values-production.yaml (hunk 5 above)
Comment (artemistomaras) [reply stable/rabbitmq/values-production.yaml:259]: Will fix this

**Comment by artemistomaras** [comment]
@juan131 I implemented the requested changes

File: stable/rabbitmq/templates/statefulset.yaml [hunk 7]
```diff
@@ -53,6 +53,20 @@ spec:
...
+        image: "{{ template "volumePermissions.image" . }}"
+        imagePullPolicy: {{ default "" .Values.volumePermissions.image.pullPolicy | quote }}
+        command: ["/bin/chown", "-R", "{{ .Values.securityContext.runAsUser }}:{{ .Values.securityContext.fsGroup }}", "{{ .Values.persistence.path }}"]
+        securityContext:
+          runAsUser: 0
+        resources:
+{{ toYaml .Values.volumePermissions.resources | indent 10 }}
+        volumeMounts:
+        - name: data
+          mountPath: "{{ .Values.persistence.path }}"
```
Comment (juan131) [inline stable/rabbitmq/templates/statefulset.yaml:63]: Could you update the `volumeMount` available in the RabbitMQ container too?

File: stable/rabbitmq/values-production.yaml (hunk 3 above)
Comment (juan131) [reply stable/rabbitmq/values-production.yaml:112]: Yes... I think it's a bug introduced on this commit: https://github.com/helm/charts/commit/8dc228412f2e63877bacb0bc5502855aaa87bcbc
Could you update the README.md so it uses `port` instead of `amqpPort`?

**Comment by artemistomaras** [comment]
@juan131 changes have been implemented. Please review them

File: stable/rabbitmq/values-production.yaml [hunk 8]
```diff
@@ -96,12 +96,13 @@ service:
   ## Node port
   ## ref: https://github.com/bitnami/bitnami-docker-rabbitmq#environment-variables
   ##
-  nodePort: 5672
+  nodePort: 30672
```
Comment (juan131) [inline stable/rabbitmq/values-production.yaml]: We should not be setting a default value to `nodePort`. As the README.md says, the default value should be random:

> random available between 30000-32767

Comment (juan131) [reply stable/rabbitmq/values-production.yaml]: As you can check in the `values.yaml`, this parameter is commented.

**Review by juan131 — COMMENTED** [review]
Some minor change and I'm okay with the changes.

File: stable/rabbitmq/values-production.yaml (hunk 8 above)
Comment (artemistomaras) [reply stable/rabbitmq/values-production.yaml]: @juan131 that makes sense. The service is of type ``ClusterIP`` in the production values anyway. I've already made the necessary changes.

**Review by juan131 — APPROVED** [review]
I'm approving your PR, sorry for the late response! Thanks for implementing the suggestions. 

Please rebase your branch from master and update it so we can merge it.

**Comment by artemistomaras** [comment]
@juan131  I resolved the conflicts with master and rebase from upstream master. Please verify that everything is ok.
